### PR TITLE
Use generic attributes to pass HTML attributes to view components

### PR DESCRIPTION
**Pull request by bockets**

### CODE REVIEW

**Review by mgallagher0 — APPROVED** [review]

File: _patterns/using-html-attributes-with-view-components.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+# Good
+
+```ruby
+class MyViewComponent < ApplicationViewComponent
+  attr_reader :variant, :attributes
+
+  def initialize(variant:, **attributes)
+    @variant = variant
+    @attributes = attributes
```
Comment (johngallagher) [inline _patterns/using-html-attributes-with-view-components.md:43]: Hmmm... wouldn't be keen on just having any attributes at the top level as discussed above - it'd be better IMO to have it nested under a specific key.

This makes the interface more discoverable:

```ruby
def initialize(variant:, html:)
  # ...
end
```

It also keeps all HTML attributes together neatly:

### Harder to parse

```ruby
Form.new(id: "5678", variant: Variant.new, class: "form")
```

vs

### Easier to parse

```ruby
Form.new(variant: Variant.new, html: { id: "5678", class: "form" })
```

File: _patterns/using-html-attributes-with-view-components.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  end
+end
+```
+
+```erb
+<!-- the template -->
+<%= content_tag(:div, class: "some-class #{Array(attributes[:class]).join(" ")}", **attributes.except(:class)) do %>
```
Comment (RenatoCoelho-BiggerPockets) [inline _patterns/using-html-attributes-with-view-components.md:50]: It's not the most readable code at first glance, I would have that class not class filtering on helper methods.
```suggestion
<%= content_tag(:div, class: "some-class #{dynamic_classes}", **html_attributes) do %>
```

File: _patterns/using-html-attributes-with-view-components.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+---
+categories: Rails
+name: "Using HTML Attributes in View Components"
+---
+
+When authoring a view component, in general you want to be able for the result to be used in ways that are already supported by the browser. For that reason, each ViewComponent initializer method should take a generic keyword argument named "attributes" that is passed through to the view as the HTML attributes. Why? This keeps our ViewComponent implementations small and flexible.
```
[the remark this answers is not in the document]
Comment (bockets) [reply _patterns/using-html-attributes-with-view-components.md:6]: Good idea. I have updated the example.

File: _patterns/using-html-attributes-with-view-components.md (hunk 1 above)
Comment (bockets) [reply _patterns/using-html-attributes-with-view-components.md]: Agreed!

File: _patterns/using-html-attributes-with-view-components.md (hunk 2 above)
Comment (bockets) [reply _patterns/using-html-attributes-with-view-components.md]: I updated the example based on @johngallagher's recommendation. I agree, it makes more sense than everything at the top level.

**Comment by bockets** [comment]
Closing because this breaks down when considering the case that a component has multiple top-level tags, such as a component that renders a `label` and `input` tags.

**Pull request closed without merging**